components/resource_control/src/future.rs

```diff
@@ -44,3 +124,312 @@ impl<F: Future> Future for ControlledFuture<F> {
         res
     }
 }
+
+// `TrackIoStateFuture` is used to update the thread local variable
+// `THREAD_TOTAL_IO_BYTES`. it should be used pair with `LimitedFuture` so the
+// `THREAD_TOTAL_IO_BYTES` state is valid.
+#[pin_project]
+pub struct TrackIoStateFuture<F> {
+    #[pin]
+    f: F,
+}
+
+impl<F: Future> TrackIoStateFuture<F> {
+    #[allow(dead_code)]
+    pub fn new(f: F) -> Self {
+        Self { f }
+    }
+}
+
+impl<F: Future> Future for TrackIoStateFuture<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        THREAD_TOTAL_IO_BYTES.with(|s| {
+            let mut stat = s.get();
+            stat.state = LocalIoState::MaybeOutdated;
+            s.set(stat);
+        });
+        let res = this.f.poll(cx);
+        THREAD_TOTAL_IO_BYTES.with(|s| {
+            let mut stat = s.get();
+            if stat.state == LocalIoState::MaybeOutdated {
+                stat.state = LocalIoState::Outdated;
+                s.set(stat);
+            }
+        });
+        res
+    }
+}
+
+// `LimitedFuture` wraps a Future with ResourceLimiter, it will automically
+// statistics the cpu time and io bytes consumed by the future, and do async
+// waiting according the configuration of the ResourceLimiter.
+#[pin_project]
+pub struct LimitedFuture<F: Future> {
+    #[pin]
+    f: F,
+    #[pin]
+    pre_delay: OptionalFuture<Compat01As03<Delay>>,
+    #[pin]
+    post_delay: OptionalFuture<Compat01As03<Delay>>,
+    resource_limiter: Arc<ResourceLimiter>,
+    res: Poll<F::Output>,
+}
+
+impl<F: Future> LimitedFuture<F> {
+    #[allow(dead_code)]
+    pub fn new(f: F, resource_limiter: Arc<ResourceLimiter>) -> Self {
+        Self {
+            f,
+            pre_delay: None.into(),
+            post_delay: None.into(),
+            resource_limiter,
+            res: Poll::Pending,
+        }
+    }
+}
+
+impl<F: Future> Future for LimitedFuture<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+        if !this.post_delay.is_done() {
+            assert!(this.pre_delay.is_done());
+            std::mem::swap(&mut *this.pre_delay, &mut *this.post_delay);
```

Comment:
why need `pre_delay`, can we only use the `post_delay`?

Reply:
because a future can only be polled once in each round, this is a restriction of the rust compiler.